src/canvas.py

```diff
@@ -104,11 +106,99 @@ def __init__(self, application):
         self._legend = True
         self._legend_position = misc.LEGEND_POSITIONS[0]
         self._handles = []
+        self.mpl_connect("scroll_event", self.on_scroll_event)
+        self.mpl_connect("motion_notify_event", self.get_coord_fraction)
+        self.xfrac = 0
+        self.yfrac = 0
+        zoom_gesture = Gtk.GestureZoom.new()
+        zoom_gesture.connect("scale-changed", self.on_zoom_gesture)
+        self.add_controller(zoom_gesture)
 
     def get_application(self):
         """Get application property."""
         return self.props.application
 
+    def get_coord_fraction(self, event):
+        xlim = self.top_right_axis.get_xlim()
+        ylim = self.top_right_axis.get_ylim()
+        figure_settings = \
+            self.get_application().get_data().get_figure_settings()
+        xscale = figure_settings.get_top_scale()
+        yscale = figure_settings.get_right_scale()
+
+        if event.inaxes is not None:
+            self.xfrac = utilities.get_fraction_at_value(
+                event.xdata, xlim[0], xlim[1], xscale)
+            self.yfrac = utilities.get_fraction_at_value(
+                event.ydata, ylim[0], ylim[1], yscale)
+
+    def on_zoom_gesture(self, _gesture, scale):
+        scale = 1 + 0.15 * (scale - 1)
+        if scale > 5 or scale < 0.2:
+            # Don't scale if ridiculous values are registered
+            scale = 1
+        self.on_zoom_event(scale)
+
+    def on_scroll_event(self, event):
+        scale = 1.6
+        if event.button == "up":
+            scale = 1 / scale
+        if self.get_application().get_ctrl() is True:
+            self.on_zoom_event(scale)
+
+    def on_zoom_event(self, scaling=1.6):
+        for ax in [self.axis]:
```

Comment:
shouldn't this be applied to all axes?

Reply:
Thanks, yes I put this to only the first axis for testing purposes. Forgot to bring it back